### PR TITLE
SDK V2

### PR DESCRIPTION
**Pull request by clydeshaffer**
# SDK V2 #

I'm grouping together a bunch of improvements to the SDK, some of which are ported with tweaks from Ganymede. Overall I feel it should merit a major version number because it will change a lot of API names and semantics. The goal of these new features is to automate away some pain points as well as better communicate system behaviors by the kinds and names of functions provided. A common theme with these improvements is to allow for larger-scale projects where you can work on a game for longer without it becoming an unmaintainable tangle.

## New sprite allocation system ##
 `allocate_sprite` selects an open quadrant from the 32 (8x4) quadrants of Sprite RAM pages. It returns a `SpriteSlot` handle that can be passed to functions that draw sprites. The import scripts now generate Load Lists which are linked lists representing how a sprite sheet is intended to be loaded and whether it covers multiple quadrants. To load a sprite, its Load List is passed to `allocate_sprite`, which will attempt to fit it into available Sprite RAM space.
 
 ## Aseprite Animation Tag Macros ##
 The import scripts will now generate a `.json.h` file for the sprite packing data exported from Aseprite, which includes the start/end frame numbers for each animation tag as well as their length. This is used to refer to animations by name instead of number, so that if an sprite's animations get reorganized or change length you only need to run `make import` to update the frame values.

## Reorganized Graphics API ##
`drawing_funcs` has been split into `gfx_sys`, `draw_queue`, and `draw_direct`. Queued and non-queued draw calls have been separated into different headers, and renamed with `queue_` and `direct_` prefixes to make it clear which calls belong to which API. The Direct Drawing API has added setup functions to prepare for drawing sprites or boxes in Direct Mode. There has also been an added direct version of draw_sprite_frame.

Redundant text drawing code has also been removed.

## Reorganized Audio API ##
`DynaWave` is not a very descriptive name, and refers to a specific hardware prototype that emulated by the first versions of the ACP firmware. APIs that interact with the Audio Coprocessor or its firmware now refer to it as `audio_coprocessor` instead of `DynaWave`

## Project config JSON ##
`project.json` has been added to the project root to config project-wide settings like the ROM name and number of PROG banks. PROG banks are extra banks added that aren't associated to an assets folder. Bank numbers for PROGx banks are surfaced in a generated header file `bank_nums.h` as macros eg. `BANK_PROG0`

## Optional SDK Features ##
Certain features like the default Random Number Generator, Text Renderer, or Flash Saving code are now gated behind macros that can be controlled from the new `project.json`. If a module isn't listed in the config file then it won't even be compiled. Features are on by default.

Current optional features:
+ RANDOM - default RNG implementation
+ TEXT - default text renderer
+ PERSIST - API for persisting data to cartridge flash memory
+ DRAWQUEUE - The queued drawing system and its API

## Asset Index Tables ##
The import scripts now generate tables of the asset pointers and bank numbers divided by the file extension of the asset. Eg. `.bmp` files get a separate table and indexing from `.mid` and `.sfx`. Previously generated ways to refer to this data will continue to exist. These tables are linked into the LOADERS segment like the sprite loading linked lists so they won't make you run out of room in the boot bank. The purpose of these tables is to be able to have certain API functions take an index instead of a pointer and a bank, reducing the number of bytes passed as function arguments by 2.

## Sound Effects on All Channels ##
Previously sound effects were hardcoded to play on channel 2, now they can be played independently of each other on any of the four channels. `play_sound_effect` has been altered to use the index mentioned above, and sound effect files should use the `.sfx` extension. The `priority` parameter has been replaced with `channel` which indicates the channel to play the sound effect on, but can still be used to give a priority number. For example: `play_sound_effect(ASSET__snd__bell_low_sfx_ID, 2 | SFX_PRIORITY(3));`

## Sound Effect Instrument Mode ##
Instrument number 0 now sets Percussion Mode on a music channel, where the MIDI note number is used as an index into the SFX table and the unmodified SFX sequence will be played on that channel with priority 0. The MIDI note number should be the intended index plus one, due to note zero having the special meaning of a "Note Off" event.

## ROM Banking Changes ##
The functionality of `change_rom_bank` has been split into `change_rom_bank` and `push_rom_bank`. `change_rom_bank` still takes a char bank number as its sole argument but no longer saves the current bank to the bank stack. `push_rom_bank` with no arguments should be called before `change_rom_bank`. This allows calling `change_rom_bank` multiple times without wasting time performing redundant bank changes. Typically a function that changes the bank should call `push_` at the beginning, then `change_` as many times as is needed, and `pop_` before returning.

## Splash Screen ##
Added a logo screen that says "GAMETANK" and the sort of dunDUN! sound you might expect from a console booting up.
Also means that the audio and graphics will already be initialized when main() is called, so you don't need to call init_graphics, init_audio_coprocessor, and init_music directly.

### CODE REVIEW

File: scripts/build_setup/build_setup.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+`
+);
+
+fs.writeFileSync(modulesHeaderFileName, 
+    `//@generated
+//Editing this manually is not recommended
+//This file tells the SDK optional features which ones are enabled
+${projectConfig.modules.map((mod) => `#define ENABLE_MODULE_${mod}`).join('\n')}
+`
+);
```
Comment (nickgirardo) [inline scripts/build_setup/build_setup.js:55]: I don't really like how optional features are enabled/ disabled by custom syntax in this json file. I do appreciate that this approach does succeed in keeping the C and asm synced, as if that wasn't a concern just having users manually edit their `modules_enabled.h` would be much nicer (imo).

I struggled for a bit to think of an approach that wouldn't require either keeping a c defs file and and asm defs file in sync or a bespoke format to produce generated files.  On thing that came to mind was defining something in the makefile like `ENABLED_MODULES = ...` and then using something like `sed` to prepare it to be appended to `CFLAGS` and `AFLAGS` e.g.:

```
ENABLED_MODULES ?= "PERSIST RANDOM TEXT"

CFLAGS += $(shell echo $(ENABLED_MODULES) | sed s/[^" "]*/-DENABLE_MODULE_\&/g)
AFLAGS += $(shell echo $(ENABLED_MODULES) | sed s/[^" "]*/-D" "ENABLE_MODULE_\&/g)
```

Although this solution feels similarly janky and ad hoc.  I'm curious to what your thoughts are

Comment (nickgirardo) [reply scripts/build_setup/build_setup.js:55]: Nit: if you do decide to keep the above, I think `//This file describes which optional features of the SDK are enabled` would be a bit clearer

**Comment by nickgirardo** [comment]
A bit of a nit, but you've included `scripts/build_setup/node_modules` in the tree.  These files should ideally not be checked into vc

**Comment by nickgirardo** [comment]
Apologies as I'm typing this without first testing it, but it appears as though only the function bodies are removed from the build.  This is obviously nice as it's where the bytes on the cartridge come from, but I think this might make difficult to debug errors if users have misconfigured projects.

If I do not have (for instance) random enabled and in `main.c` I mistakenly include `random.h` and call `rnd()`, I expect that this would be a difficult to diagnose linker error. From the perspective of `main.c`, `rnd()` should exist but when `random.c` is compiled it doesn't output the symbol.

It should be possible to start the header files corresponding to the disabled features with
```
#ifndef ENABLE_MODULE_RANDOM
#error "Module random is not enabled! Please check *location* to make sure your modules are correctly configured"
#endif
```

This would cause a compilation error when compiling `random.c` which currently imports `random.h` unconditionally, so including `random.h` would have to be done conditionally based on whether `ENABLE_MODULE_RANDOM` is set

File: scripts/build_setup/build_setup.js (hunk 1 above)
Comment (nickgirardo) [reply scripts/build_setup/build_setup.js:55]: This above solution would also mean that it isn't necessary to include `modules_enabled.h` or `modules_enabled.inc` in each of the optional modules

Comment (clydeshaffer) [reply scripts/build_setup/build_setup.js:55]: One thing about the solution that uses the `modules_enabled.h` file is that VSCode knows easily whether to gray out the syntax between the ifdefs, without having to get it to understand flags passed to cc65 in the makefile.

NodeJS projects keep these kind of configs in a `package.json` fairly often, and I think one particular thing it does well is separating configuration from the build mechanics. At work I'm constantly told "oh you just change that in so-and-so file" and then have to go swim through a thirty page makefile and eventually decode which was meant to be the configurable part. The `project.json` here is meant to be a file that only contains things it is okay to edit.

**Comment by clydeshaffer** [comment]
Updated the SDK documentation for V2
https://wiki.gametank.zone/doku.php?id=development:csdk:2.0:csdk

**Pull request merged**